File: frontend/services/auditLogService.ts
import api from '@/lib/axios';

export type AuditAction = 'CREATE' | 'UPDATE' | 'DELETE' | 'LOGIN' | 'EXPORT';

export interface AuditLog {
  id: number;
  action: AuditAction;
  entityName: string;
  entityId?: string | null;
  userId?: number | null;
  username?: string | null;
  userRole?: string | null;
  ipAddress?: string | null;
  userAgent?: string | null;
  reason?: string | null;
  oldValues?: Record<string, unknown> | null;
  newValues?: Record<string, unknown> | null;
  createdAt: string;
}

export interface AuditLogFilters {
  userId?: number;
  username?: string;
  action?: AuditAction | '';
  entityName?: string;
  userRole?: string;
  fromDate?: string;
  toDate?: string;
}

export interface AuditLogFilterOptions {
  actions: string[];
  entityNames: string[];
  usernames: string[];
  roles: string[];
}

export interface PageResponse<T> {
  content: T[];
  totalPages: number;
  totalElements: number;
  number: number;
  size: number;
  first: boolean;
  last: boolean;
}

export const auditLogService = {
  getLogs: async (
    page: number = 0,
    size: number = 20,
    filters?: AuditLogFilters,
  ) => {
    const params: Record<string, string | number> = { page, size };

    if (filters) {
      if (filters.userId) params.userId = filters.userId;
      if (filters.username) params.username = filters.username;
      if (filters.action) params.action = filters.action;
      if (filters.entityName) params.entityName = filters.entityName;
      if (filters.userRole) params.userRole = filters.userRole;
      if (filters.fromDate) params.fromDate = filters.fromDate;
      if (filters.toDate) params.toDate = filters.toDate;
    }

    const response = await api.get('/admin/audit-logs', { params });
    return response.data as { data: PageResponse<AuditLog> };
  },

  getById: async (id: number) => {
    const response = await api.get(`/admin/audit-logs/${id}`);
    return response.data as { data: AuditLog };
  },

  getFilterOptions: async (): Promise<AuditLogFilterOptions> => {
    const response = await api.get('/admin/audit-logs/filters');
    return response.data?.data;
  },
};

File: backend/src/main/java/com/cinema/service/AuditLogService.java
package com.cinema.service;

import com.cinema.audit.AuditAction;
import com.cinema.dto.response.AuditLogFilterOptionsResponse;
import com.cinema.dto.response.AuditLogResponse;
import com.cinema.model.AuditLog;
import com.cinema.repository.AuditLogRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.time.LocalDateTime;
import java.util.Objects;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AuditLogService {

    private final AuditLogRepository auditLogRepository;

    public Page<AuditLogResponse> getAuditLogs(
            Long userId,
            String username,
            AuditAction action,
            String entityName,
            String userRole,
            LocalDateTime fromDate,
            LocalDateTime toDate,
            int page,
            int size) {
        Pageable pageable = PageRequest.of(
                Math.max(page, 0),
                Math.min(Math.max(size, 1), 100),
                Sort.by(Sort.Direction.DESC, "createdAt"));

        Specification<AuditLog> specification = Specification.where(null);

        if (userId != null) {
            specification = specification.and((root, query, cb) -> cb.equal(root.get("userId"), userId));
        }
        if (StringUtils.hasText(username)) {
            String pattern = "%" + username.trim().toLowerCase() + "%";
            specification = specification.and((root, query, cb) -> cb.like(cb.lower(root.get("username")), pattern));
        }
        if (action != null) {
            specification = specification.and((root, query, cb) -> cb.equal(root.get("action"), action));
        }
        if (StringUtils.hasText(entityName)) {
            specification = specification.and((root, query, cb) -> cb.equal(root.get("entityName"), entityName.trim()));
        }
        if (StringUtils.hasText(userRole)) {
            specification = specification.and((root, query, cb) -> cb.equal(root.get("userRole"), userRole.trim()));
        }
        if (fromDate != null) {
            specification = specification.and((root, query, cb) -> cb.greaterThanOrEqualTo(root.get("createdAt"), fromDate));
        }
        if (toDate != null) {
            specification = specification.and((root, query, cb) -> cb.lessThanOrEqualTo(root.get("createdAt"), toDate));
        }

        return auditLogRepository.findAll(specification, pageable).map(this::toResponse);
    }

    public AuditLogResponse getAuditLogById(Long id) {
        AuditLog auditLog = auditLogRepository.findById(id)
                .orElseThrow(() -> new com.cinema.exception.ResourceNotFoundException("AuditLog", "id", id));
        return toResponse(auditLog);
    }

    public AuditLogFilterOptionsResponse getFilterOptions() {
        return AuditLogFilterOptionsResponse.builder()
                .actions(auditLogRepository.findDistinctActions().stream()
                        .filter(Objects::nonNull)
                        .map(Enum::name)
                        .distinct()
                        .sorted()
                        .toList())
                .entityNames(auditLogRepository.findDistinctEntityNames())
                .usernames(auditLogRepository.findDistinctUsernames())
                .roles(auditLogRepository.findDistinctUserRoles())
                .build();
    }

    private AuditLogResponse toResponse(AuditLog auditLog) {
        return AuditLogResponse.builder()
                .id(auditLog.getId())
                .action(auditLog.getAction())
                .entityName(auditLog.getEntityName())
                .entityId(auditLog.getEntityId())
                .userId(auditLog.getUserId())
                .username(auditLog.getUsername())
                .userRole(auditLog.getUserRole())
                .ipAddress(auditLog.getIpAddress())
                .userAgent(auditLog.getUserAgent())
                .reason(auditLog.getReason())
                .oldValues(auditLog.getOldValues())
                .newValues(auditLog.getNewValues())
                .createdAt(auditLog.getCreatedAt())
                .build();
    }
}
